// thirdweb/dist/cjs/rpc/fetch-rpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchSingleRpc = exports.fetchRpc = void 0;
const fetch_js_1 = require("../utils/fetch.js");
const json_js_1 = require("../utils/json.js");
/**
 * @internal
 */
async function fetchRpc(rpcUrl, client, options) {
    const response = await (0, fetch_js_1.getClientFetch)(client)(rpcUrl, {
        headers: {
            ...client.config?.rpc?.fetch?.headers,
            "Content-Type": "application/json",
        },
        body: (0, json_js_1.stringify)(options.requests),
        method: "POST",
        requestTimeoutMs: options.requestTimeoutMs ?? client.config?.rpc?.fetch?.requestTimeoutMs,
        keepalive: client.config?.rpc?.fetch?.keepalive,
    });
    if (!response.ok) {
        response.body?.cancel();
        throw new Error(`RPC request failed with status ${response.status} - ${response.statusText}`);
    }
    let result;
    if (response.headers.get("Content-Type")?.startsWith("application/json")) {
        result = await response.json();
    }
    else {
        result = await response.text();
    }
    return result;
}
exports.fetchRpc = fetchRpc;
/**
 * @internal
 */
async function fetchSingleRpc(rpcUrl, client, options) {
    const response = await (0, fetch_js_1.getClientFetch)(client)(rpcUrl, {
        headers: {
            ...(client.config?.rpc?.fetch?.headers || {}),
            "Content-Type": "application/json",
        },
        body: (0, json_js_1.stringify)(options.request),
        method: "POST",
        requestTimeoutMs: options.requestTimeoutMs ?? client.config?.rpc?.fetch?.requestTimeoutMs,
        keepalive: client.config?.rpc?.fetch?.keepalive,
    });
    if (!response.ok) {
        response.body?.cancel();
        throw new Error(`RPC request failed with status ${response.status}`);
    }
    let result;
    if (response.headers.get("Content-Type")?.startsWith("application/json")) {
        result = await response.json();
    }
    else {
        result = await response.text();
    }
    return result;
}
exports.fetchSingleRpc = fetchSingleRpc;
