// thirdweb/dist/esm/rpc/fetch-rpc.js
import { getClientFetch } from "../utils/fetch.js";
import { stringify } from "../utils/json.js";
/**
 * @internal
 */
export async function fetchRpc(rpcUrl, client, options) {
    const response = await getClientFetch(client)(rpcUrl, {
        headers: {
            ...client.config?.rpc?.fetch?.headers,
            "Content-Type": "application/json",
        },
        body: stringify(options.requests),
        method: "POST",
        requestTimeoutMs: options.requestTimeoutMs ?? client.config?.rpc?.fetch?.requestTimeoutMs,
        keepalive: client.config?.rpc?.fetch?.keepalive,
    });
    if (!response.ok) {
        response.body?.cancel();
        throw new Error(`RPC request failed with status ${response.status} - ${response.statusText}`);
    }
    let result;
    if (response.headers.get("Content-Type")?.startsWith("application/json")) {
        result = await response.json();
    }
    else {
        result = await response.text();
    }
    return result;
}
/**
 * @internal
 */
export async function fetchSingleRpc(rpcUrl, client, options) {
    const response = await getClientFetch(client)(rpcUrl, {
        headers: {
            ...(client.config?.rpc?.fetch?.headers || {}),
            "Content-Type": "application/json",
        },
        body: stringify(options.request),
        method: "POST",
        requestTimeoutMs: options.requestTimeoutMs ?? client.config?.rpc?.fetch?.requestTimeoutMs,
        keepalive: client.config?.rpc?.fetch?.keepalive,
    });
    if (!response.ok) {
        response.body?.cancel();
        throw new Error(`RPC request failed with status ${response.status}`);
    }
    let result;
    if (response.headers.get("Content-Type")?.startsWith("application/json")) {
        result = await response.json();
    }
    else {
        result = await response.text();
    }
    return result;
}
